### PR TITLE
Add cnvway GMTK API to Segway

### PR DESCRIPTION
**Pull request by isaiahah**

### CODE REVIEW

**Comment by EricR86** [comment]
The `input_master.py` should just be renamed to `gmtk.py` and live in the `segway` folder. As it stands the current proposed `input_master.py` would not be importable since the gmtk directory is missing a `__init__.py` to mark it as a module. Is there a reason for the subdirectory and/or gmtk submodule that I'm missing?

File: doc/api.rst [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+    appropriate attributes to describe the model.
+
+    2. A line saving the :py:class:`InputMaster` object to an input file, 
+    which describes the parameters in GMTK structure format. 
+
+    3. Code calling Segway to train the defined model and annotate using the 
+    trained model. More information on running Segway is available in 
+    :ref:`python-interface`.
+
+The CNVway code provides a worked example applying the GMTK API to defining,
```
Comment (EricR86) [inline doc/api.rst:37]: Where would this be? Maybe just link to github repo?

File: doc/api.rst [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+Usage example:
+
+.. code-block:: python
+
+    # Create InputMaster object
+    input_master = InputMaster()
+    # Set parameters
+    ...
+    # Save to output file
+    input_master.save(outfile)
```
Comment (EricR86) [inline doc/api.rst:153]: probably rename `outfile` to something more filename-ish instead of a potential file handle.

File: doc/api.rst [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+    # Create a Covar object in the InputMaster covar InlineSection
+    input_master.covar["dist1"] = Covar(1.0)
+    # Alternately, a numeric value will be converted to a Covar
+    input_master.covar["dist2"] = 1.0
+
+
+.. py:class:: DPMF
+
+    A Numpy ``ndarray`` representing a DPMF with a specialized string method
```
Comment (EricR86) [inline doc/api.rst:235]: Should define the acronym DPMF

**Comment by isaiahah** [comment]
You previously recommended changing the import statement in Figure 1 to `from segway.gmtk.input_master import ...` so I created the gmtk directory as an attempt to replicate that; I forgot the `__init__.py` to mark it as a module. However, I didn't have any other reason for placing the file there, so unless you think of anything I will move it to `segway/gmtk.py`.

File: doc/api.rst [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+
+    # Create a custom DPMF object in the InputMaster mean InlineSection
+    input_master.dpmf["biased"] = DPMF([0.7, 0.3])
+    # Create a uniform DPMF with a specified shape
+    input_master.dpmf["uniform"] = DPMF.uniform_from_shape(3)
+
+
+.. py:class:: DiagGaussianMC
+
+    A Gaussian distribution with diagonal covariance. Currently the only
```
Comment (EricR86) [inline doc/api.rst:267]: Is this correct? I thought it was referring to the diagonal components in a matrix. Also please define and expand any acronyms.

File: doc/api.rst [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+    # Create a MX objects in the InputMaster mx InlineMXSection.
+    # Arguments are labels for DPMF and MX objects.
+    input_master.mx["emission1"] = MX("biased", ["dist1", "dist2"])
+    input_master.mx["emission2"] = MX("biased", ["dist1", "dist2"])
+
+
+.. py:class:: DenseCPT
+
+    A Numpy ``ndarray`` representing a dense CPT with a specialized string method
```
Comment (EricR86) [inline doc/api.rst:342]: Again expand the CPT acronym at least once and anywhere else in the docs for other acronyms that show up.

File: doc/api.rst [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+        :type cardinality: int
+        :param dt: Name of an existing decision tree 
+        :type dt: str 
+
+
+Section Classes
+===============
+
+Classes to store multiple objects that form one section of the parameter file.
+These are used within the ``InputMaster`` object and need not be defined by
```
Comment (EricR86) [inline doc/api.rst:416]: This is strange. Why are these documented then if no one who reads it is supposed to use it? I thought these would be useful in some weird unimplemented GMTK object corner cases?

**Review by EricR86 — CHANGES_REQUESTED** [review]
@isaiahah let me know when you've addressed these comments and I'll do another quick round before passing it on.

**Comment by michaelmhoffman** [comment]
Let's discuss what the file will be called

Michael M. Hoffman, PhD
Senior Scientist and Chair, Computational Biology Medicine Program
  Princess Margaret Cancer Centre, University Health Network
Associate Professor, Departments of Medical Biophysics and Computer Science
  University of Toronto
Faculty Affiliate, Vector Institute for Artificial Intelligence

Princess Margaret Cancer Research Tower 11-311
101 College St
Toronto, ON M5G 1L7

https://hoffmanlab.org/

File: doc/api.rst (hunk 4 above)
Comment (isaiahah) [reply doc/api.rst]: Clarified this refers to a diagonal covariance matrix.

File: doc/api.rst (hunk 6 above)
Comment (isaiahah) [reply doc/api.rst:421]: I'm unsure if these should be documented, as casual users will not need them but anyone extending the software will. I will discuss it at the meeting on Friday.

**Comment by isaiahah** [comment]
I've renamed `input_master.py` as discussed in the meeting. As we discussed I've also added an Internal Classes section to the GMTK API documentation. Please review that section and let me know if I'm including too much detail, too little detail, or if there's other internal parts of the API that I missed documenting.

File: segway/gmtk/input_master.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+        """
+        # Ensure all arguments are strings or lists of strings
+        if not all(isinstance(name, str) for name in names):
+            raise TypeError("All arguments must be str instances.")
+
+        super().__init__(names)
+
+    def __str__(self) -> str:
+        """
+        Returns string format of NameCollection object to be printed into the
```
Comment (michaelmhoffman) [inline segway/gmtk/input_master.py:172]: Imperative mood

File: segway/gmtk/input_master.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+    return cls(value)
+
+
+class DPMF(OneLineKind):
+    """
+    A single DPMF object.
+    """
+    kind = OBJ_KIND_DPMF
+
+    # todo check if sums to 1.0
```
Comment (michaelmhoffman) [inline segway/gmtk/input_master.py:259]: -> `TODO` or `XXX`

Comment (michaelmhoffman) [reply segway/gmtk/input_master.py:259]: Please either post an issue or pull request for this.

**Review by michaelmhoffman — COMMENTED** [review]
A couple of minor changes. This is approved after those changes made

File: segway/gmtk/input_master.py (hunk 8 above)
Comment (isaiahah) [reply segway/gmtk/input_master.py:259]: I'll resolve this change as part of the GMTK API integration branch.

**Comment by isaiahah** [comment]
CI tests are failing in the install step due to an updated Cython (3.03, from 3.02) requiring a newer version of blocs2 (2.2.8, from 2.0.0), which requires Python>=3.9 when only Python 3.8.18 is installed as part of the test. Eric is changing CI to use Python 3.9; that change will be merged in, then this branch can be merged to main.

**Pull request merged**